xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
+    '''
+    Base class that implements utilities for PromptFormatters.
 
-    # step out completely if git is not installed
-    try:
-        binary_location = subprocess.check_output(['which', 'git'], cwd=cwd,
-                                    stderr=subprocess.PIPE,
-                                    universal_newlines=True)
-        if not binary_location:
-            return branch
-    except subprocess.CalledProcessError:
-        return branch
+    You should read the docs for this class to learn how to write your own
+    Prompt Formatter.  However, you most likely want to inherit from
+    DefaultPromptFormatter if you are writing a custom formatter.
+
+    Custom Prompt Formatters will need to implement their own __init__()
+    method that adds entries to several dicts depending on the needs of the
+    variables they add.  The Formatter may also implement methods to implement
+    those variables (although very simple variables can be written with
+    pre-existing functions.)
+
+    Example:
```

Comment:
Thanks for these questions @abadger 

> And I'm not sure if xonsh has configured sphinx to pull documentation from the class's docstring or the __init___ docstring (When I've used sphinx in the past it's been an either or).

I believe that I have it set up to use both.  It appends the `__init__` docstring to the class docstring.

> I got to the point where I had to modify the Makefile to use sphinx-build-3.4 so that it could use autodoc on the source.  

You should not need to do that.

> but then it appeared that cloud_sptheme wasn't ported to python3 yet. 

It does wok on Python 3. I have used it many times. See https://pythonhosted.org/cloud_sptheme/install.html#requirements

> (pip3.4 install --user cloud_sptheme fails). Any clues on how you do it so that I can be sure the documentation is all building correctly?

I don't know what is going on here, and you might report it to them.  I used conda to get the package, not pip.  If pip is failing though, you can always try from source.  See the previous link.  Also, it is in a lot of package managers, depending on what OS you are on.